File: cracking_the_coding_interview_6th/chapter17/python/17_08_v02.py
from typing import List


def find_max_height(hw: List[List[int]]) -> int:
    hw = sorted(hw)
    vs = []
    res = 0
    for i in range(len(hw)):
        v = []
        for j in range(0, i):
            if vs[j][-1] < hw[i][1] and len(v) < len(vs[j]):
                v = vs[j]
        vs.append(v[:] + [hw[i][1]])
        res = max(res, len(vs[i]))
    return res


def ns(f):
    return next(f).strip()


def solve(fp: str):
    print(f"# {fp}")
    with open(fp) as f:
        n = int(ns(f))
        hw = [list(map(int, ns(f).split())) for _ in range(n)]
    print(find_max_height(hw))
    print()


solve("../testcases/17_08/01.txt")
solve("../testcases/17_08/02.txt")
solve("../testcases/17_08/03.txt")

# # ../testcases/17_08/01.txt
# 6

# # ../testcases/17_08/02.txt
# 6

# # ../testcases/17_08/03.txt
# 1

File: cracking_the_coding_interview_6th/chapter17/cpp/17_08_v02.cpp
#include <algorithm>
#include <fstream>
#include <string>
#include <vector>
using namespace std;
using ll = long long;
using pll = pair<ll, ll>;

ll find_max_height(vector<pll> hw) {
  sort(hw.begin(), hw.end());
  vector<vector<ll>> vs;
  ll res = 0;
  for (ll i = 0; i < hw.size(); ++i) {
    vector<ll> _v;
    vector<ll> *v = &_v;
    for (ll j = 0; j < i; ++j) {
      if (*(vs[j].end() - 1) < hw[i].second && v->size() < vs[j].size()) {
        v = &(vs[j]);
      }
    }
    vs.emplace_back(v->begin(), v->end());
    vs[i].push_back(hw[i].second);
    res = max(res, (ll)vs[i].size());
  }
  return res;
}

void solve(const string &fp) {
  printf("# %s\n", fp.c_str());
  ifstream ifs{fp};
  ll n;
  ifs >> n;
  vector<pll> hw(n);
  for (ll i = 0; i < n; ++i) {
    ll h, w;
    ifs >> h >> w;
    hw[i] = {h, w};
  }
  printf("%lld\n", find_max_height(hw));
  puts("");
}

int main() {
  solve("../testcases/17_08/01.txt");
  solve("../testcases/17_08/02.txt");
  solve("../testcases/17_08/03.txt");
}

// # ../testcases/17_08/01.txt
// 6

// # ../testcases/17_08/02.txt
// 6

// # ../testcases/17_08/03.txt
// 1